### DISCUSSION TITLE
OpenGlControlBase does not detect the mouse pointer anymore

### DISCUSSION THREAD

**Discussion by gsanhueza** [AvaloniaUI/Avalonia#10879] [Q&A] [answered]
Hello!
I wanted to study how OpenGL works in the upcoming version of Avalonia, so I took an available example in https://github.com/kekekeks/Avalonia-Silk.NET-Example

In particular, I'm interested in the `PointerEntered` event.
The `PointerEntered` event is triggered correctly when using the version described in that example (`11.0.999-cibuild0027451-beta`).

When I updated the version from `11.0.999-cibuild0027451-beta` to `11.0.0-preview4` in the _.csproj_ file of the example, the `PointerEntered` event is also triggered correctly.

In both cases, it's possible to verify it by using the Avalonia DevTools (F12): When passing the mouse above the OpenGL control, what appears in the "Logical Tree" is:

```
>Application
  >MainWindow (:pointerover)
    >SilkNetExample (:pointerover)
```

But then I updated the version from `11.0.0-preview4` to `11.0.0-preview5`, replacing the bare minimum (because of the [breaking changes](https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes)):

- Added the `Avalonia.Themes.Fluent` package and removed `Mode="Light"` from "FluentTheme" in `App.xaml.cs`
- Replaced `OnOpenGlInit(GlInterface gl, int fb)` with `OnOpenGlInit(GlInterface gl)` (because the API requires it)
- Replaced `OnOpenGlDeinit(GlInterface gl, int fb)` with `OnOpenGlDeinit(GlInterface gl)` (because the API requires it too)

With the update and these changes, the `PointerEntered` event is not triggered anymore, and what appears in Avalonia DevTools (F12) when passing the mouse above the OpenGL control is:

```
>Application
  >MainWindow (:pointerover)
    >SilkNetExample
```

Updating from `11.0.0-preview5` to `11.0.0-preview6` does not fix the issue.

As I see it, the change from `11.0.0-preview4` to `11.0.0-preview5` broke `OpenGlControlBase`'s mouse detection, which means it's not possible to use the control to get the `PointerEntered` event anymore.

I also tested the `PointerMoved`, `PointerPressed` and `PointerWheelChanged` events and they're also broken, which I assume it's because the `OpenGlControlBase` control is not detecting when the mouse enters/leaves.

**Comment by gsanhueza** [comment]
I studied the behavior of other controls that do show `:pointerover` in the _Logical Tree_ of Avalonia DevTools when putting the mouse pointer above the control.

For example, one of the controls (Canvas), did not show `:pointerover`, which I found strange.
Researching more, I found [this issue](https://github.com/AvaloniaUI/Avalonia/issues/5276) which exposed a similar "bug", but it was solved by using `Background=Transparent` in the XAML file.

For example:

```
<Grid RowDefinitions="*,*">
    <!-- This canvas does not react to the mouse pointer -->
    <Canvas Grid.Row="0" />

    <!-- But this canvas does! -->
    <Canvas Background="Transparent" Grid.Row="1" />
</Grid>
```

If I connect this idea to my original issue, that means the "broken" `OpenGlControlBase` control is not broken: it seems to behave that way intentionally.

https://user-images.githubusercontent.com/19938125/229402828-1828c90b-7f5b-4427-bc04-1388fb91adbd.mp4


But I still need to detect a mouse when it enters the OpenGL control, because I intend to react to that, by overriding the following methods:

```
protected override void OnPointerEntered(PointerEventArgs e) => Console.WriteLine("PointerEntered");
protected override void OnPointerExited(PointerEventArgs e) => Console.WriteLine("PointerExited");
```

Unfortunately, I can't just use `Background="Transparent"` with an `OpenGlControl` that inherits from `OpenGlControlBase` to make it work, because the base control does not have a `Background` property.

Thus, the question is: How can I get a control inheriting from `OpenGlControlBase` to react to mouse inputs, so my overridden methods can work again, as they were working in the `11.0.0-preview4` version?

**Reply by timunie** [reply]
As there were a lot of bug fixes since preview4, can you maybe double-check if it works in latest master for you?

**Reply by gsanhueza** [reply]
At the moment I can't run Avalonia from source (maybe I'm missing something?)

For the record, I'm testing Avalonia using Windows 11 and Git Bash.

I followed the instructions found in https://github.com/AvaloniaUI/Avalonia/blob/master/Documentation/build.md

I executed the following commands:

- `git clone https://github.com/AvaloniaUI/Avalonia.git`
- `cd Avalonia`
- `git submodule update --init`
- `cd samples\ControlCatalog.NetCore`
- `dotnet restore`
- `dotnet run`

At the time of writing this response, the `master` branch is at commit `ec6cd48de9408cb39b12f52cf4af708a93e42988`.

But the build failed, which means I cannot test at the moment whether the latest commits fix `OpenGlControlBase`'s mouse pointer issues.

Am I missing something, or is `master` not usable at this particular moment?

<img width="713" alt="image" src="https://user-images.githubusercontent.com/19938125/229546274-eca48308-335c-4ff4-b9e6-5efa3688fd9d.png">

**Reply by gsanhueza** [reply]
@timunie It seems my machine had something wrong, because when I rebooted, I was able to execute the ControlCatalog at the `master` branch (commit `ec6cd48de9408cb39b12f52cf4af708a93e42988`).

Nonetheless, the `OpenGlPageControl` in ControlCatalog still can't detect when the mouse pointer passes above it, but the rest of the controls in that page are able to detect it.

Is there something I can do to restore the `PointerEntered`/`PointerExited` functionality, like the Canvas example above, but this time for `OpenGlControlBase`?

https://user-images.githubusercontent.com/19938125/229867238-47da3637-5a76-4b2f-af25-c1938986541c.mp4

**Comment by kekekeks** [comment, marked as answer]
You can implement ICustomHitTest as a workaround.

**Reply by gsanhueza** [reply]
The workaround works! I added the following code in the ControlCatalog example to test it, and now the `PointerEntered`/`PointerExited` functionality is restored.

```
public class OpenGlPageControl : OpenGlControlBase, ICustomHitTest
{
    // [Original code of the ControlCatalog example]

    bool ICustomHitTest.HitTest(Point point) => true;
}

```

With the addition of the workaround, now all relevant events also work, such as `PointerPressed`, `PointerReleased`, `PointerMoved`, `PointerWheelChanged` and `PointerCaptureLost`.

I also tested what would happen if the workaround was integrated into Avalonia code:

```
public abstract class OpenGlControlBase : Control, ICustomHitTest
{
    // [Original code of Avalonia.OpenGL.Controls.OpenGlControlBase]

    // Workaround added to Avalonia code
    bool ICustomHitTest.HitTest(Point point) => true;
}
```

Plus:

```
public class OpenGlPageControl : OpenGlControlBase
{
    // [Original code of the ControlCatalog example, without the workaround]
}

```

And the OpenGL example of ControlCatalog also works, now without requiring the workaround (because it's integrated in the `OpenGlControlBase` code).

I do not know if the workaround can be integrated that directly (because the `HitTest` method only returns `true`), but if it's reasonable to add it before releasing Avalonia 11.0, that would be awesome!

And if it's not possible, the workaround still works to restore the missing functionality, and it's now documented in this discussion.

Thanks a lot!

**Reply by vpenades** [reply]
I am using a variation of this hack:

```c#
 bool ICustomHitTest.HitTest(Point point)
 {  
     return this.Bounds.Contains(point);           
 }
 ```

this ensures the hack only affects the openGL panel and not everything else.

**Reply by NeverMorewd** [reply]
I would like share my workaround

```csharp
public bool HitTest(Point point) => new Rect(Bounds.Size).Contains(point);
```
